feat: save secrets on storage create

ciyer opened the pull request:
Allow storing credentials when a storage is created.

# Screenshots

Add an option to save credentials to the last page of the add data source flow

<img width="814" alt="image" src="https://github.com/user-attachments/assets/5180b588-991d-438d-b76a-f6d6b4b63d43">

When checked, the credentials will be stored after the data source is added. This can be verified by looking at the data source, the credential field will be displayed as <saved secret>

<img width="493" alt="image" src="https://github.com/user-attachments/assets/58b8c266-520b-4857-b081-bc38af9940b2">


# Testing

Add a data source that requires credentials to a project (or example an S3 or polybox data source). Check that the credentials used to test are stored. Also, check that there is no option to store credentials for a public data source (like a public S3 bucket).


/deploy renku-notebooks=master renku-data-services=main

andre-code reviewed (changes requested):
The code looks good overall, but there are a few issues that need to be addressed:

1. The secret title is currently displayed as the ID, but this ID is not shown in the data storage information. It would be better to use a more descriptive name, such as project/data-storageName.

![Screenshot 2024-08-26 at 13 23 40](https://github.com/user-attachments/assets/fb7b8a7f-29dd-4f94-879d-bd406ab1fef7)

2. When a secret is added while launching the store, it is saved as a user secret but doesn't show in the user secret list and does not indicate that it is saved in the storage information as `<saved credential>`.



3. The icons are not properly aligned. I’ve left a comment in the code regarding this issue.

4. The data storage sidebar should not open when clicking on the form label while editing a data storage.
![save credentials B](https://github.com/user-attachments/assets/f713d653-3c49-41a6-8b21-a74b380ad936)

![no show credentials saved](https://github.com/user-attachments/assets/bc82056c-9212-4ff9-843d-d334dea899b7)

5. Clicking on user secrets redirects to Renku 1.0. It would be better to have a Renku 2.0 route for user secrets when accessed from Renku 2.0.

5. Here would be helpful to confirm that the secret is also created successfully.

![Screenshot 2024-08-26 at 13 22 49](https://github.com/user-attachments/assets/73a3e4dc-0313-4736-ae08-ac26781e21f4)

6. The checkbox to store the secret is not shown when creating the data source, although the secret is still stored.

Please address these issues here or create new issues to handle them in a separate PR

ciyer commented on client/src/features/project/components/cloudStorage/AddOrEditCloudStorage.tsx:
```diff
@@ -1259,23 +1291,70 @@ function AddStorageMount({
+      {storage.storageId == null &&
+        isV2 &&
+        hasPasswordFieldWithInput &&
+        validationSucceeded && (
```
Fixed in 959724cf4799da74e8d6085985ebb75737e2f554

ciyer replied:
Fixed by not saving the credentials if the field is not shown in 65dd221a85a3f286945fc1224549f14d64fc10a2

ciyer commented on client/src/features/ProjectPageV2/ProjectPageContent/DataSources/DataSourceDisplay.tsx:
```diff
@@ -152,17 +152,17 @@ export function DataSourceActions({
           data-cy="data-source-credentials"
           onClick={toggleCredentials}
         >
           <Lock className={cx("bi", "ms-1")} />
```
Fixed in 959724cf4799da74e8d6085985ebb75737e2f554

ciyer commented on client/src/features/project/components/cloudStorage/AddOrEditCloudStorage.tsx:
```diff
@@ -1259,23 +1291,70 @@ function AddStorageMount({
+  state: AddCloudStorageState;
+};
+
+function AddStorageMountSaveCredentialsInfo({
```
Fixed in 756da48ffa7055c1e1b44015e2fd7ffe774afe00

ciyer commented on client/src/features/project/components/cloudStorage/projectCloudStorage.constants.ts:
```diff
@@ -138,6 +138,7 @@ export const EMPTY_CLOUD_STORAGE_STATE: AddCloudStorageState = {
   showAllSchema: false,
   showAllProviders: false,
   showAllOptions: false,
+  saveCredentials: true,
```
Fixed by not saving the credentials if the field is not shown in 65dd221a85a3f286945fc1224549f14d64fc10a2

ciyer commented:
Thanks for the review!


> 2. When a secret is added while launching the store, it is saved as a user secret but doesn't show in the user secret list and does not indicate that it is saved in the storage information as `<saved credential>`.
> 3. The icons are not properly aligned. I’ve left a comment in the code regarding this issue.
> 7. [After skipping the credentials test, the] checkbox to store the secret is not shown when creating the data source, although the secret is still stored.

These bugs have been fixed.

> 6. Here would be helpful to confirm that the secret is also created successfully.

I added this information to the dialog box

![image](https://github.com/user-attachments/assets/40ac29e5-1fed-482c-b9eb-28bf49da9b2b)

> 4. The data storage sidebar should not open when clicking on the form label while editing a data storage.

You are right -- I was able to reproduce this bug, but it has been around for longer, and was not introduced by the changes here. I will address it in a separate PR.

> 
> 1. The secret title is currently displayed as the ID, but this ID is not shown in the data storage information. It would be better to use a more descriptive name, such as project/data-storageName.
>
> 5. Clicking on user secrets redirects to Renku 1.0. It would be better to have a Renku 2.0 route for user secrets when accessed from Renku 2.0.

I agree, the current display of the secrets is not really usable. These limitations, however, were deemed acceptable in realizing the pitch, and improving the display will require changes to the backend because there is no efficient way to get the data source that a secret is for, so this will need to be handled separately, with support from the backend.

> 
> Please address these issues here or create new issues to handle them in a separate PR

andre-code commented:
Thanks for applying those fixes!

I've noticed one issue and a couple of minor CSS adjustments that need attention:

[Bug] If the storage secret is removed from the secret page, the status on the project page doesn't update. As a result, when launching the session, the missing credential is not required.

![no ask credentials](https://github.com/user-attachments/assets/3e5ca82f-f45b-46a0-bc0e-63afee51fa24)

[CSS] In the storage credential modal, there's no vertical padding between fields. Please add some padding.

![Screenshot 2024-09-02 at 13 39 11](https://github.com/user-attachments/assets/ba39bc36-2b50-4f44-8ba5-ef155e202719)


[CSS] Similarly, on the data source page, there's no vertical padding between "Required credentials" and "Access mode." Please address this as well.

![Screenshot 2024-09-02 at 13 39 22](https://github.com/user-attachments/assets/8f016bac-34c5-4576-ad71-38f63c8a94af)

ciyer commented:
Thanks for the review, I have fixed the issues you have raised.

> [Bug] If the storage secret is removed from the secret page, the status on the project page doesn't update. As a result, when launching the session, the missing credential is not required.

Fixed in 311382d6f5f16a016026b8e9e5886dbfcfad467f

> [CSS] In the storage credential modal, there's no vertical padding between fields. Please add some padding.

![image](https://github.com/user-attachments/assets/3f15e84d-6431-460b-b7ce-7b7caeeb8db8)

Fixed in 65ce6e475a475faf80f5c9c31e85ee3b85103bd0

> [CSS] Similarly, on the data source page, there's no vertical padding between "Required credentials" and "Access mode." Please address this as well.

![image](https://github.com/user-attachments/assets/6299840d-c344-4d41-b828-72951b905df2)

Fixed in 212c0c5db1dddd4cc274c2977a7233c55841689c

andre-code reviewed (approved):
lgtm 🚀
